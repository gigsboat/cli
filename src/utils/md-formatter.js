import json2md from 'json2md'
import { getConfig } from '../../bin/config-manager.js'

json2md.converters.rawHTML = function (input) {
  return input
}

export function formatToMarkdown(data) {
  return json2md(data)
}

export async function getEventsMd(events) {
  let markdownEventsContent = ''
  for (const yearlyItems of events) {
    const eventsByYear = await eventsListForYear(yearlyItems)
    markdownEventsContent += eventsByYear + '\n'
  }

  const markdown = getTableOfContents(events) + '\n' + markdownEventsContent
  return markdown
}

function getTableOfContents(events) {
  const tableOfContents = []
  tableOfContents.push({
    h1: 'Table of Contents'
  })

  const markdownYearsItems = []
  for (const yearlyItems of events) {
    markdownYearsItems.push(
      `[Year of ${yearlyItems.year}](#${yearlyItems.year}) - total events ${yearlyItems.stats.total}`
    )
  }

  tableOfContents.push({
    ul: markdownYearsItems
  })

  return json2md(tableOfContents)
}

export function getStatsBadges(events) {
  const statsBadges = []

  const badgeForTotalEvents = events.stats.total
    ? `![Total Events](https://img.shields.io/badge/total-${events.stats.total}-blue?style=flat-square)`
    : 0
  const badgeForTotalConferences = events.stats.total_conference
    ? `![Total Conferences](https://img.shields.io/badge/conferences-${events.stats.total_conference}-red?style=flat-square)`
    : 0
  const badgeForTotalPodcasts = events.stats.total_podcast
    ? `![Total Podcasts](https://img.shields.io/badge/podcasts-${events.stats.total_podcast}-yellow?style=flat-square)`
    : 0
  const badgeForTotalWebinars = events.stats.total_webinar
    ? `![Total Webinars](https://img.shields.io/badge/webinars-${events.stats.total_webinar}-lightgrey?style=flat-square)`
    : 0
  const badgeForTotalMeetups = events.stats.total_meetup
    ? `![Total Meetups](https://img.shields.io/badge/meetups-${events.stats.total_meetup}-violet?style=flat-square)`
    : 0
  const badgeForTotalArticles = events.stats.total_article
    ? `![Total Podcasts](https://img.shields.io/badge/articles-${events.stats.total_article}-green?style=flat-square)`
    : 0
  const badgeForTotalWorkshops = events.stats.total_workshop
    ? `![Total Workshops](https://img.shields.io/badge/workshops-${events.stats.total_workshop}-orange?style=flat-square)`
    : 0

  statsBadges.push({
    p: `${badgeForTotalEvents ? badgeForTotalEvents : ''} ${
      badgeForTotalMeetups ? badgeForTotalMeetups : ''
    } ${badgeForTotalConferences ? badgeForTotalConferences : ''} ${
      badgeForTotalPodcasts ? badgeForTotalPodcasts : ''
    } ${badgeForTotalWebinars ? badgeForTotalWebinars : ''} ${
      badgeForTotalArticles ? badgeForTotalArticles : ''
    } ${badgeForTotalWorkshops ? badgeForTotalWorkshops : ''}`
  })

  return json2md(statsBadges) + '\n'
}

async function eventsListForYear(eventsOfYear) {
  const gigsConfig = await getConfig()

  const eventsByYear = []
  let eventsImageGalleryItems = []

  eventsByYear.push({
    h1: eventsOfYear.year
  })
  eventsByYear.push(getStatsBadges(eventsOfYear))

  const eventsTableEntries = []
  eventsOfYear.items.forEach((event) => {
    const eventDate = `${event.attributes.date.getUTCFullYear()}-${
      event.attributes.date.getMonth() + 1
    }-${event.attributes.date.getDate()}`

    const pathToFileOnGitHub = encodeURI(event.fileRelativePath)

    eventsTableEntries.push({
      Date: eventDate,
      Event: event.attributes.name,
      Title: event.fileRelativePath
        ? `[${event.attributes.title}](${pathToFileOnGitHub})`
        : event.attributes.title,
      Slides: event.attributes.slides_url
        ? `[Slides](${event.attributes.slides_url})`
        : '',
      Recording: event.attributes.recording_url
        ? `[Recording](${event.attributes.recording_url})`
        : '',
      Location: event.attributes.country_code
        ? event.attributes.country_code
        : '',
      Language: event.attributes.language ? event.attributes.language : ''
    })

    if (event.attributes.images) {
      const imagesList = event.attributes.images.filter((item) => !!item)
      eventsImageGalleryItems = eventsImageGalleryItems.concat(imagesList)
    }
  })

  if (gigsConfig.output.includePictureGalleryYearly === true) {
    let tableItems = []
    let tableHTML = ''

    for (let i = 0; i <= eventsImageGalleryItems.length - 1; i + 7) {
      tableItems.push(eventsImageGalleryItems.splice(i, 8))
    }

    if (tableItems.length) {
      tableHTML += '<table>' + '\n'
      tableItems.forEach((itemsRow) => {
        if (itemsRow.length) {
          tableHTML += '  <tr>' + '\n'
          itemsRow.forEach((item) => {
            tableHTML +=
              `    <td align="center"> <img src="${item}" width="85" height="50" /> </td>` +
              '\n'
          })
          tableHTML += '  </tr>' + '\n'
        }
      })
      tableHTML += '</table>' + '\n'
    }

    eventsByYear.push(json2md({ rawHTML: tableHTML }))
  }

  eventsByYear.push({
    table: {
      headers: [
        'Date',
        'Event',
        'Title',
        'Slides',
        'Recording',
        'Location',
        'Language'
      ],
      rows: eventsTableEntries
    }
  })

  return json2md(eventsByYear) + '\n'
}
