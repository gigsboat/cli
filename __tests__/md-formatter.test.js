import { formatToMarkdown, getEventsMd } from '../src/utils/md-formatter.js'

test('convert json object to markdown formatting', async () => {
  const json = [
    {
      h1: 'great title'
    },
    {
      p: 'and my paragraph is cool'
    }
  ]
  const markdown = formatToMarkdown(json)
  expect(markdown).toMatchSnapshot()
})

test('process events json data to provide a complete markdown document', async () => {
  const events = [
    {
      year: 2016,
      items: [
        {
          attributes: {
            date: new Date('2016-01-01'),
            title: 'title',
            name: 'name',
            slides_url: 'https://example.com',
            recording_url: 'https://example.com',
            country_code: 'US',
            language: null
          }
        }
      ],
      stats: {
        total: 1
      }
    },
    {
      year: 2021,
      items: [
        {
          attributes: {
            date: new Date('2021-02-01'),
            title: 'title',
            type: 'conference',
            name: 'name',
            slides_url: 'https://example.com',
            recording_url: null,
            country_code: null,
            language: 'English'
          }
        },
        {
          attributes: {
            date: new Date('2021-02-02'),
            title: 'title',
            type: 'meetup',
            name: 'name',
            slides_url: null,
            recording_url: 'https://example.com',
            country_code: 'US',
            language: 'English'
          }
        },
        {
          attributes: {
            date: new Date('2021-02-02'),
            title: 'title',
            type: 'podcast',
            name: 'name',
            slides_url: null,
            recording_url: 'https://example.com',
            country_code: 'US',
            language: 'English'
          }
        },
        {
          attributes: {
            date: new Date('2021-02-02'),
            title: 'title',
            type: 'article',
            name: 'name',
            slides_url: null,
            recording_url: 'https://example.com',
            country_code: 'US',
            language: 'English'
          }
        },
        {
          attributes: {
            date: new Date('2021-02-02'),
            title: 'title',
            type: 'webinar',
            name: 'name',
            slides_url: null,
            recording_url: 'https://example.com',
            country_code: 'US',
            language: 'English'
          }
        },
        {
          attributes: {
            date: new Date('2021-02-02'),
            title: 'title',
            type: 'webinar',
            name: 'name',
            slides_url: null,
            recording_url: 'https://example.com',
            country_code: 'US',
            language: 'English'
          }
        },
        {
          attributes: {
            date: new Date('2021-02-02'),
            title: 'title',
            type: 'workshop',
            name: 'name',
            slides_url: null,
            recording_url: 'https://example.com',
            country_code: 'US',
            language: 'English'
          }
        },
        {
          attributes: {
            date: new Date('2021-02-03'),
            title: 'title',
            type: 'workshop',
            name: 'name',
            slides_url: null,
            recording_url: 'https://example.com',
            country_code: 'US',
            language: 'English'
          }
        }
      ],
      stats: {
        total: 8,
        total_podcast: 1,
        total_conference: 1,
        total_webinar: 2,
        total_meetup: 1,
        total_article: 1,
        total_workshop: 2,
        total_other: 0
      }
    }
  ]

  const markdown = await getEventsMd(events)
  expect(markdown).toMatchSnapshot()
})
