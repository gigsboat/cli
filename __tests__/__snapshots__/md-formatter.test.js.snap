// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert json object to markdown formatting 1`] = `
"# great title


and my paragraph is cool
"
`;

exports[`process events json data to provide a complete markdown document 1`] = `
"# Table of Contents


 - [Year of 2016](#2016) - total events 1
 - [Year of 2021](#2021) - total events 8

# 2016


![Total Events](https://img.shields.io/badge/total-1-blue?style=flat-square)      


| Date | Event | Title | Slides | Recording | Location | Language |
| ---- | ----- | ----- | ------ | --------- | -------- | -------- |
| 2016-1-1 | name | title | [Slides](https://example.com) | [Recording](https://example.com) | US |  |


# 2021


![Total Events](https://img.shields.io/badge/total-8-blue?style=flat-square) ![Total Meetups](https://img.shields.io/badge/meetups-1-violet?style=flat-square) ![Total Conferences](https://img.shields.io/badge/conferences-1-red?style=flat-square) ![Total Podcasts](https://img.shields.io/badge/podcasts-1-yellow?style=flat-square) ![Total Webinars](https://img.shields.io/badge/webinars-2-lightgrey?style=flat-square) ![Total Podcasts](https://img.shields.io/badge/articles-1-green?style=flat-square) ![Total Workshops](https://img.shields.io/badge/workshops-2-orange?style=flat-square)


| Date | Event | Title | Slides | Recording | Location | Language |
| ---- | ----- | ----- | ------ | --------- | -------- | -------- |
| 2021-2-1 | name | title | [Slides](https://example.com) |  |  | English |
| 2021-2-2 | name | title |  | [Recording](https://example.com) | US | English |
| 2021-2-2 | name | title |  | [Recording](https://example.com) | US | English |
| 2021-2-2 | name | title |  | [Recording](https://example.com) | US | English |
| 2021-2-2 | name | title |  | [Recording](https://example.com) | US | English |
| 2021-2-2 | name | title |  | [Recording](https://example.com) | US | English |
| 2021-2-2 | name | title |  | [Recording](https://example.com) | US | English |
| 2021-2-3 | name | title |  | [Recording](https://example.com) | US | English |


"
`;
